Project/DAQ_System/Program/Src/Sensor/Accelerometer/lsm303dlhc.cpp

```diff
@@ -0,0 +1,171 @@
+
+    calibrate();
+}
+
+void LSM303DLHC::calibrate() {
+    static constexpr uint8_t kSampleCount = 100;
+    double acc_magnitude = 0;
+    double average_array[kSampleCount];
+
+    // Collects samples to calibrate for 1G
+    for(int i = 0; i < kSampleCount; i++) {
+        ReadRawAcceleration();
+        acc_magnitude = sqrt(raw_acceleration_data_[0] * raw_acceleration_data_[0] + raw_acceleration_data_[1] * raw_acceleration_data_[1] + raw_acceleration_data_[2] * raw_acceleration_data_[2]);
+        average_array[i] = acc_magnitude * 0.001;
+    }
+
+    double min_value = std::numeric_limits<double>::max();
+    double max_value = -1;
+
+    // Find the largest and smallest magnitudes
+    for(int j = 2; j < kSampleCount; j++) {
+        if(average_array[j] <= min_value) {
+            min_value = average_array[j];
+
+        } else if (average_array[j] > max_value) {
+            max_value = average_array[j];
+        }
+    }
+
+    // Check that the maximum deviation is below a certain threshold.
+    // If it is too large, the conversion factor will be set to a previously
+    // determined conversion factor
+    gravity_adjustment_conversion_factor_ = 0;
+
+    if (max_value - min_value < 0.09) {
+        double total_array = 0;
+	    for(int i = 0; i < kSampleCount; i++) {
+		    total_array += average_array[i];
+	    }
+
+	    gravity_adjustment_conversion_factor_ = total_array / kSampleCount;
+    } else {
+        // Testing found that ~1.028 is about 1G
+        gravity_adjustment_conversion_factor_ = 1.028;
+    }
+}
+
+void LSM303DLHC::SetScale(float x, float y, float z) {
+    scale_[0] = x;
+    scale_[1] = y;
+    scale_[2] = z;
+}
+
+void LSM303DLHC::SetOffset(float x, float y, float z) {
+    offset_[0] = x;
+    offset_[1] = y;
+    offset_[2] = z;
+}
+
+float* LSM303DLHC::GetAcceleration() {
+    return real_acceleration_data_;
+}
+
+void LSM303DLHC::ComputeAcceleration() {
+    ReadRawAcceleration();
+/*   old calculations for accelerometer
+    static constexpr float kMG_LSB = .001F;
+    real_acceleration_data_[0] = (raw_acceleration_data_[0]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+    real_acceleration_data_[1] = (raw_acceleration_data_[1]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+    real_acceleration_data_[2] = (raw_acceleration_data_[2]) * kMG_LSB / gravity_adjustment_conversion_factor_;
+*/
+
+    // linearization of the accelerometer. by default range of the accelerometer is ±2 g
+    real_acceleration_data_[0] = 0.0006 *raw_acceleration_data_[0] - 0.0002;
+    real_acceleration_data_[1] = 0.0006 *raw_acceleration_data_[1] - 0.0002;
+    real_acceleration_data_[2] = 0.0006 *raw_acceleration_data_[2] - 0.0002;
+}
+
+void LSM303DLHC::ReadRawAcceleration() {
+    uint8_t command[1] = { OUT_X_L_A | 0x80 };
+    HAL_I2C_Master_Transmit(&i2c_,ACC_ADDRESS, command,1, HAL_MAX_DELAY);
+
+    uint8_t bytes_received[6];
+    HAL_I2C_Master_Receive(&i2c_,ACC_ADDRESS, bytes_received,6, HAL_MAX_DELAY);
```

Comment:
Add constants for the amount of bytes to send and receive — for the sake of less "magic numbers".